gammapy/cube/basic_cube.py

```diff
@@ -0,0 +1,62 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""Functions to perform basic functions for map and cube analysis.
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+from ..maps import WcsNDMap
+
+def make_map_counts(evts, valid_ref_map):
+    """
+    Build a ``WcsNDMap` with events from an EventList.
+    The energy of the events is used for a non-spatial axis homogeneous to energy.
+    The other non-spatial axis names should have an entry in the colum names of the ``EventList``
+
+    Parameters
+    __________
+
+    evts : `~gammapy.data.EventList`
+            the input event list
+    valid_ref_map : `~gammapy.maps.WcsNDMap`
+        Map containing the valid pixels
+
+    Returns
+    -------
+    cntmap : `~gammapy.maps.WcsNDMap`
+        Count cube (3D) in true energy bins
+
+    """
+    # The list will contain the event table entries to be fed into the WcsNDMap
+    tmp = list()
+    ref_geom = valid_ref_map.geom
+    # Convert events coordinates
+    if ref_geom.coordsys == 'GAL':
+        tmp.append(evts.galactic.l)
+        tmp.append(evts.galactic.b)
+    elif ref_geom.coordsys == 'CEL':
+        tmp.append(evts.radec.ra)
+        tmp.append(evts.radec.dec)
+    else:
+        # should raise an error here. The map is not correct.
+        raise ValueError("Incorrect coordsys of input map.")
+
+    for i, axis in enumerate(ref_geom.axes):
+        if axis.unit.is_equivalent("eV"):
+            # This axis is the energy
+            tmp.append(evts.energy.to(axis.unit))
+        elif axis.name.upper() in evts.table.colnames:
+            # Here we assume that colanmes are all capital
+            tmp.append(evts.table[axis.name.upper()].to(axis.unit))
+        elif axis.name.lower() in evts.table.colnames:
+            # Here we assume that colanmes are not capital
+            tmp.append(evts.table[axis.name.lower()].to(axis.unit))
+        else:
+            raise ValueError("Cannot find MapGeom axis {} in EventList", axis.name)
+
+    # Create map
+    cntmap = WcsNDMap(ref_geom)
+    # Fill it
+    cntmap.fill_by_coords(tmp)
+
+    # Put counts outside validity region to zero
```

Comment:
I'm -1 on doing the mask here. A function should do one thing. The caller can just do a second line `apply_mask` if needed. What if I don't have or want a mask?

If you get rid of this, you can also simplify the docstring / signature.
Should this take a ref_geom?
Or maybe just take a map and fill it?
That would be more efficient, when filling one map from many event lists, no, avoiding the temp maps for each run?

Reply:
I think this a map maker. It should be applied to a single observation to create the count map not to fill a larger map with different observations.
One can add a fill_map function that takes a ``WcsNDMap`` and an ``EventList`` and use it there. 
The mask can applied outside. But it needs to be applied to each map produced (background, exposure, etc). It seems simpler to apply it in each maker. Again one can make another function apply_mask to be called by each maker.

Reply:
Agreed.

But IMO it should go like this for the case if the mask is different for each obs:
```
geo = <your choice>
counts_map = Map.empty(geo)
for obs in observations:
    map_fill_counts(counts_map, obs.events)
    map_apply_mask(counts_map, obs.mask)
```
Probably @woodmd has thoughts on the question of whether to have that code in one or two functions, and which one should be a map method or a standalone function, and whether the map data is modified in-place vs copy?

Reply:
I agree that separate methods for filling and masking will be more flexible.  I also think a `mask` map method would make more sense than a standalone function.  I would envision a method that accepts a geometry object and zeros pixels outside the geomtry (e.g. as determined by calling `contains`).

Reply:
In addition we need the case of just multiplying with a pre-computed mask. For the maximum FOV offset cut, we want to compute the mask only once and then apply it to counts, exposure, background. It should be possible to do that without computing the mask 3x, even if it's not expensive. It could be just a one-liner method, or a little more that also gives good error messages when mask to apply and map don't match.

@registerrier - since you'll be offline next week, we should merge this PR in before. Please do any edits you have time for, and then I'll either take over and continue work here next week, or just merge and re-start from master.

Reply:
Note that the logic here is rather the following:
geo = <your choice>
count_map = Map.empty(geo)
expo_map = Map.empty(geo)
for obs in observations:
    obs_geom = geo.cutout(some FoVcut)
    obs_counts_map = Map.empty(obs_geom) 
    map_fill_counts(obs_counts_map, obs.events)
    map_apply_mask(obs_counts_map, obs.mask)
    count_map.coadd(obs_count_map)
    
    obs_expo_map = Map.empty(obs_geom)
    compute_exposure(obs_expo_map,  aeff, livetime)
    map_apply_mask(obs_expo_map, obs.mask)
    expo_map.coadd(obs_expo_map)
   ...

So we are not simply applying a fill_map, we have to make a cutout and define a mask for that specific observation. Also the map creation step is repeated for exposure and background.
I think we can expose ``fill_map_counts`` and ``map_apply_mask`` but having a maker function that creates the the map, fills it and apply a mask (if there's one) is simpler.
The mask is not computed is the make function. It is an argument.

Reply:
@registerrier - but note that the current function is still not in line with having modular code, i.e. the "single responsibility" principle that a function or class should do one thing.

You have the part inside that does the generic events to map axis matching and handling, then the fill, then the mask apply. We can have functions of classes whose job it is to orchestrate these multiple steps, that's fine. But the individual steps should be available in separate functions.

But as I said: just finish up here as you like / have time, and then we'll just iterate and refactor ...

Reply:
OK. This is done. 
Where should the masking be performed? Directly in maps: ``Map.apply_mask(mask)``?